### PR TITLE
Feat/layers laplace conv

### PR DESCRIPTION
**Pull request by sinapordanesh**
# ✨ Laplace Neural Operator (LNO) ✨
This PR introduces **Laplace‑based Spectral Convolutions** in 1‑D, 2‑D, and 3‑D, along with unit tests and a small utility for shape enforcement.

**Primary references**

| Resource | Link |
|----------|------|
| Paper | <https://arxiv.org/pdf/2303.10528> |
| Original implementation | <https://github.com/qianyingcao/Laplace-Neural-Operator> |

---

## 1  |  Why a series of small PRs?
Implementing Laplace and Wavelet kernels touches multiple layers of the library (kernels → blocks → models).  
To keep each review focused and mergeable, I’m submitting a **bottom‑up sequence** of PRs:

1. **Spectral kernels** (Laplace ✓, Wavelet → upcoming)  
2. **Block‑level integration** (FNOBlocks / XNOBlocks)  
3. **Model‑level wiring** (FNO, TFNO, etc.)  
4. **Docs & gallery example refresh**

Each PR contains **fully‑documented code, helper utilities, unit tests, and inline typing**.  
I don’t expect immediate approval; instead, I’m committed to an iterative review process and will revise as many times as needed to meet project standards.

---

## 2  |  What’s in this PR

| File | Purpose |
|------|---------|
| `neuralop/layers/spectral_convolution_laplace.py` | New `SpectralConvLaplace{1,2,3}D` layers following core API conventions (FFT shape‑handling, `modes`, bias, complex weights). |
| `neuralop/layers/shape_enforcer.py` | Utility to ensure inverse FFT outputs match expected spatial dims (PyTorch handles this for FFT; Laplace kernels need an explicit helper). |
| `neuralop/layers/tests/test_spectral_convolution_laplace.py` | Unit tests for forward pass, mode truncation, etc. |
| `neuralop/layers/tests/test_shape_enforcer.py` | Edge‑case tests for cropping/padding logic. |

---

## 3  |  Review notes & open questions

* **API surface** – I tried to mirror `SpectralConv` naming (`in_channels`, `out_channels`, etc.), and consider all those arguemtns in each new spectral convolution classes, although not all of them being used. Just keep them for future developement. 
* **ShapeEnforcer location** – currently in `neuralop.layers`; relocate if there’s a utilities sub‑package you’d rather use.

---

## 4  |  Additional references
For a detailed comparison between the Laplace kernels introduced in this PR and the original paper’s implementation, see **LNO.md**:

[LNO.md](https://github.com/user-attachments/files/19823392/LNO.md)

The note documents each kernel, explains the adaptations made for `neuralop`, and provides accuracy checks confirming numerical parity with the reference code.

---

*Thank you for your time and feedback. I’m available to iterate until the code fits seamlessly into* **neuraloperator** *and its CI.*

### CODE REVIEW

File: neuralop/layers/tests/test_spectral_convolution_laplace.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
+# MIT License
+# Copyright (c) 2024 Saman Pordanesh
```
Comment (dhpitt) [inline neuralop/layers/tests/test_spectral_convolution_laplace.py:2]: I would remove. Authors don't own individual files in the library

File: neuralop/layers/spectral_convolution_laplace.py [hunk 2]
```diff
@@ -0,0 +1,979 @@
...
+    if len(start_points) != dim or len(end_points) != dim:
+        raise ValueError(
+            "Start points and end points must match the number of input dimensions ({dim})."
+        )
+
+    # Compute domain lengths from start and end points
+    domain_lengths = [end_points[i] - start_points[i] for i in range(dim)]
+
+    # Generate grid points for each dimension using torch.linspace
+    grid = [
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:47]: In `neuraloperator`, we provide a module for performing grid embeddings that you can pass directly into your model (`neuralop.layers.embeddings.GridEmbeddingND`)

File: neuralop/layers/spectral_convolution_laplace.py [hunk 3]
```diff
@@ -0,0 +1,979 @@
...
+
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        n_modes,
+        complex_data=False,
+        max_n_modes=None,
+        bias=True,
+        separable=False,
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:120]: A lot of these parameters seem unused.

File: neuralop/layers/spectral_convolution_laplace.py [hunk 4]
```diff
@@ -0,0 +1,979 @@
...
+from typing import List, Optional, Tuple, Union
+from .resample import resample
+from ..utils import validate_scaling_factor
+from .shape_enforcer import ShapeEnforcer
+from .base_spectral_conv import BaseSpectralConv
+
+Number = Union[int, float]
+
+
+def _compute_dt(shape, start_points: List = None, end_points: List = None):
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:14]: Better to use the provided `gridembedding` layer

File: neuralop/layers/spectral_convolution_laplace.py [hunk 5]
```diff
@@ -0,0 +1,979 @@
...
+    # Compute dt directly from the grid
+    dt_list = [(grid[i][1] - grid[i][0]).item() for i in range(dim)]
+
+    return dt_list, grid
+
+
+# ====================================
+#  Laplace layer: pole-residue operation is used to calculate the poles and residues of the output
+# ====================================
+class SpectralConvLaplace1D(BaseSpectralConv):
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:61]: As a general note, it would be much better if you could find a way to refactor this logic for a dimension-agnostic case. From checking your code, nothing jumps out at me as impossible to refactor into more abstract terms, though I may be missing something.

**Review by dhpitt — CHANGES_REQUESTED** [review]
Thanks for kicking this off, @sinapordanesh . Thanks for following our conventions for commenting style as well!

I left a preliminary review with a few notes, now passing to @JeanKossaifi . My main concerns are:

1. Is there a reason the laplace convs are implemented specifically for each dimension? Over time, this makes the code more difficult to maintain and adds class proliferation. Unless there are specific elements of the computation that can't be refactored, I would suggest refactoring so that the dimension is abstracted. 
2. I don't really get the purpose of the `ShapeEnforcer`; I don't see a corresponding module in the original LNO code, and I don't see any mentions of cropping or padding in the code or the original paper. We're concerned this may impact the method's discretization-invariance.

File: neuralop/layers/shape_enforcer.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
+import torch
+import torch.nn.functional as F
+from typing import Optional, Tuple, List
+
+
+class ShapeEnforcer(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (sinapordanesh) [reply neuralop/layers/shape_enforcer.py:6]: No, it's because of integrating with the package, not originally from the paper. 
This is a copy of what parameter `s` does at:
`x = torch.fft.ifftn(out_fft, s=mode_sizes, dim=fft_dims, norm=self.fft_norm)`

`torch.fft` makes many tasks more manageable for implementing fourier-based transformations, but as Laplace and wavelet transforms are not supported in Torch, we need to implement some features from scratch, like this feature to shape the output `x` as the input `x`.

Do you think it makes sense?

File: neuralop/layers/spectral_convolution_laplace.py (hunk 3 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py:120]: Yes, considering all you've used in the original spectral_conv.py classes, for possible future feature adding. 

Do you like them to be removed for now?

File: neuralop/layers/tests/test_shape_enforcer.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
+# MIT License
+# Copyright (c) 2024 Saman Pordanesh
+
+import pytest
+import torch
+import numpy as np
+from neuralop.layers.shape_enforcer import ShapeEnforcer
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (sinapordanesh) [reply neuralop/layers/tests/test_shape_enforcer.py:10]: Done, check it out after updating my pull request.

File: neuralop/layers/spectral_convolution_laplace.py (hunk 4 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py:15]: check this out [1c398d9](https://github.com/neuraloperator/neuraloperator/pull/584/commits/1c398d94bed117f62615096055105516db05905e)

(`GridEmbeddingND` wasn't that flexible to give me the dt directly. Would make the code more complicated.)

File: neuralop/layers/spectral_convolution_laplace.py (hunk 2 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py]: check this out [1c398d9](https://github.com/neuraloperator/neuraloperator/pull/584/commits/1c398d94bed117f62615096055105516db05905e)

(`GridEmbeddingND` wasn't that flexible to give me the dt directly. Would make the code more complicated.)

File: neuralop/layers/spectral_convolution_laplace.py (hunk 3 above)
Comment (dhpitt) [reply neuralop/layers/spectral_convolution_laplace.py:104]: Parameters should only exist if they are used in the code. If/when they are needed, we can add them

File: neuralop/layers/spectral_convolution_laplace.py [hunk 8]
```diff
@@ -0,0 +1,979 @@
...
+
+        # Transform to frequency domain and compute frequency grid
+        alpha = torch.fft.fft(x, dim=-1)
+        lambda0 = (
+            torch.fft.fftfreq(t.shape[0], dt, device=alpha.device) * 2 * np.pi * 1j
+        )
+        lambda1 = lambda0.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1)
+
+        # Extract pole and residue weights from the combined weight tensor
+        weights_pole = self.weight[:, :, :modes1].view(
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:266]: Would it be simpler to store the pole and residue weights separately?

File: neuralop/layers/spectral_convolution_laplace.py [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+        tx = shape[1].to(x.device)
+        ty = shape[2].to(x.device)
+        dtz = dt_list[0]  # this can be time dimension, instead of Z dimension
+        dtx = dt_list[1]
+        dty = dt_list[2]
+
+        # Transform to frequency domain and compute frequency grids
+        alpha = torch.fft.fftn(x, dim=[-3, -2, -1])
+        omega1 = (
+            torch.fft.fftfreq(tz.shape[0], dtz, device=alpha.device) * 2 * np.pi * 1j
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py:867]: These fftfreq operations could be chained using iterables/loops to abstract away dimensions in this part

File: neuralop/layers/tests/test_shape_enforcer.py [hunk 10]
```diff
@@ -5,103 +5,134 @@
 
 
 @pytest.mark.parametrize(
-    "test_id, start_dim, input_shape, output_shape, description",
+    "start_dim, input_shape, output_shape",
```
Comment (dhpitt) [inline neuralop/layers/tests/test_shape_enforcer.py:8]: This isn't what I meant. I mean that these tests should happen in 1d, 2d, and 3d, and you should also parameterize the index of the start/stop dimensions explicitly, then compute the proper shapes inside the tests.

File: neuralop/layers/shape_enforcer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhpitt) [reply neuralop/layers/shape_enforcer.py:6]: I see. In that case, I wonder if it might be better to implement the laplace and wavelet transforms independently, then allow users to choose their transform type (e.g. fourier vs. laplace vs. wavelet) inside the convolution. What do you think, @JeanKossaifi ?

[the remark this answers is not in the document]
Comment (sinapordanesh) [reply neuralop/layers/shape_enforcer.py:6]: > I see. In that case, I wonder if it might be better to implement the laplace and wavelet transforms independently, then allow users to choose their transform type (e.g. fourier vs. laplace vs. wavelet) inside the convolution. What do you think, @JeanKossaifi ?

Yes, that’s already how I’ve structured it. The wavelet implementation hasn’t been pushed yet. As noted in the main PR comment, I’m splitting changes into separate pull requests to keep each one focused and more manageable to review. We can finish this PR for LNO first, and I’ll introduce additional features afterward.

[the remark this answers is not in the document]
Comment (sinapordanesh) [reply neuralop/layers/shape_enforcer.py:6]: > Also, have you done any empirical testing with output scaling to confirm that the layers perform as expected?

Can you give an example of the kinds of testing you think of? A test case on FNO would help.

File: neuralop/layers/spectral_convolution_laplace.py (hunk 3 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py:104]: Ok, sure
I'll remove extra parameters here.

File: neuralop/layers/tests/test_shape_enforcer.py (hunk 10 above)
Comment (sinapordanesh) [reply neuralop/layers/tests/test_shape_enforcer.py]: This should address your point [d54a405](https://github.com/neuraloperator/neuraloperator/pull/584/commits/d54a405ef9badc41a9cbecb44ade85734ec1c5da)

File: neuralop/layers/tests/test_shape_enforcer.py [hunk 11]
```diff
@@ -5,103 +5,134 @@
...
+    if x.shape[3] < output_shape[1]:
+        assert torch.all(y[:, :, :, x.shape[3]:] == 0)
 
-    # Test case specific assertions
+
+@pytest.mark.parametrize(
+    "start_dim, input_shape, output_shape, description",
+    [
+        (2, (2, 3, 10, 15), [10, 15], "Same shape - no changes needed"),
+        (2, (2, 3, 10, 15), None, "None output shape - return input unchanged"),
```
[the remark this answers is not in the document]
Comment (sinapordanesh) [reply neuralop/layers/tests/test_shape_enforcer.py]: This should address your point [d54a405](https://github.com/neuraloperator/neuraloperator/pull/584/commits/d54a405ef9badc41a9cbecb44ade85734ec1c5da)

File: neuralop/layers/spectral_convolution_laplace.py (hunk 8 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py:266]: Let me start remastering the Laplace convolution classes to have an abstract version like the one we have for FNO. In the best case, we should try to have N-dimensional, but at least we can try to have 1- 3D on a single class. 
I'll address this point there.

File: neuralop/layers/spectral_convolution_laplace.py (hunk 5 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py:45]: Let me start remastering the Laplace convolution classes to have an abstract version like the one we have for FNO. In the best case, we should try to have N-dimensional, but at least we can try to have 1- 3D on a single class.

**Review by dhpitt — COMMENTED** [review]
These changes are great, especially the general dimension-agnostic laplace convolution. Thanks! I would suggest removing the original 1, 2, and 3-D layers and performing all your tests on the new layer. Also added a few small comments.

**Comment by dhpitt** [comment]
One note on the correctness test: A relative tolerance of 5e-1 is really high. If the outputs do not pass a lower tolerance, e.g. 1e-5 or 1e-7, this means there is a numerical error in your method.

**Comment by sinapordanesh** [comment]
That's right, but in my investigation, I found the original LNO a little bit buggy. More confident in the N-dim kernel, which follows a clear path. 
However, I'll inspect further to find what's causing these differences, and report or fix it on either side.

File: neuralop/layers/spectral_convolution_laplace.py [hunk 12]
```diff
@@ -104,15 +103,22 @@ class SpectralConvLaplace(BaseSpectralConv):
         Domain start points for each dimension. Defaults to 0.0 for all.
     linspace_endpoints : list of float, optional
         Domain end points for each dimension. Defaults to 1.0 for all.
+        
+    References
+    -----------
+    .. [1] :
+
```
Comment (dhpitt) [inline neuralop/layers/spectral_convolution_laplace.py]: nit: remove these lines

Comment (dhpitt) [reply neuralop/layers/spectral_convolution_laplace.py]: e.g. 

```
.. [1] : Author, Paper name.

**Comment by sinapordanesh** [comment]
@JeanKossaifi There was a typo in the test script, causing failures on all tests. 
It should be resolved now.

File: neuralop/layers/spectral_convolution_laplace.py [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+        Domain start points for each dimension. Defaults to 0.0 for all.
+    linspace_endpoints : list of float, optional
+        Domain end points for each dimension. Defaults to 1.0 for all.
+        
+    .. [1] : Cao, Q., Goswami, S., and Karniadakis, G. E. "LNO: Laplace Neural Operator
+        for Solving Differential Equations" (2023). arXiv preprint arXiv:2303.10528,
+        https://arxiv.org/pdf/2303.10528.
+    
+    
+
```
Comment (JeanKossaifi) [inline neuralop/layers/spectral_convolution_laplace.py]: Few new lines here.

File: neuralop/layers/tests/test_spectral_convolution_laplace.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        assert list(res.shape[2:]) == list(expected_output_shape[2:])
+    
+    else:
+        # Forward pass
+        res = conv(x)
+        
+        # Check output shape
+        assert res.shape[0] == batch_size
+        assert res.shape[1] == out_channels
+        assert list(res.shape[2:]) == [12]*dim
```
Comment (JeanKossaifi) [inline neuralop/layers/tests/test_spectral_convolution_laplace.py]: Is there any test we could run besides just checking the shape?

**Review by JeanKossaifi — COMMENTED** [review]
Thank you @sinapordanesh, great contribution!

Might be useful to run black or some other tool for pep8, etc. 
Also would be great to have some doc, perhaps an example? We can also add the new block to the API.

File: neuralop/layers/spectral_convolution_laplace.py (hunk 13 above)
Comment (sinapordanesh) [reply neuralop/layers/spectral_convolution_laplace.py]: Added a brief documentation from the abstract of the LNO paper. Let me know if you'd like to see more detailed docs.

File: neuralop/layers/tests/test_spectral_convolution_laplace.py (hunk 14 above)
Comment (sinapordanesh) [reply neuralop/layers/tests/test_spectral_convolution_laplace.py]: Thanks for reminding me of this. I added a few more test assertions to cover more class features. 
Tried to match what you've implemented in test_spectral_conv.py
The LNO class has fewer arguments to be tested, and mainly the output shape. 

I've added weights complex value assurance, output real value, and testing domain description for each dimension on the LNO initialisation (important as we need to specify the domain for each problem, based on the physical domain)

**Comment by sinapordanesh** [comment]
> Might be useful to run black or some other tool for pep8, etc.
> Also would be great to have some doc, perhaps an example? We can also add the new block to the API.

For pep8 style, ran black command for all four files contained in this PR. 

Regarding block for LNO and WNO, and main classes like LNO.py, we discussed that it's a better practice first to merge spectral convolutions, then add blocks and main classes as separate PRs, to avoid complex PRs.
I'll implement those as soon as we finalise the accepted versions of each spectral convolution.

**Comment by aaprasad** [comment]
Is this going to be merged any time soon? Would love to try it out!
